src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1771,22 +1864,103 @@ public sealed class One : Tests
                     builder.CheckConsistentDecision(
                         test: test,
                         other: Test,
+                        state: state,
                         whenTrueValues: whenTrueValues,
                         whenFalseValues: whenFalseValues,
                         syntax: test.Syntax,
+                        lengthTest: out Tests lengthTest,
                         trueTestPermitsTrueOther: out bool trueDecisionPermitsTrueOther,
                         falseTestPermitsTrueOther: out bool falseDecisionPermitsTrueOther,
                         trueTestImpliesTrueOther: out bool trueDecisionImpliesTrueOther,
                         falseTestImpliesTrueOther: out bool falseDecisionImpliesTrueOther,
                         foundExplicitNullTest: ref foundExplicitNullTest);
-                    whenTrue = trueDecisionImpliesTrueOther ? Tests.True.Instance : trueDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
-                    whenFalse = falseDecisionImpliesTrueOther ? Tests.True.Instance : falseDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
+                    whenTrue = trueDecisionImpliesTrueOther ? makeOrSequence(lengthTest, this) : trueDecisionPermitsTrueOther ? this : makeAndSequence(Not.Create(lengthTest), this);
+                    whenFalse = falseDecisionImpliesTrueOther ? makeOrSequence(lengthTest, this) : falseDecisionPermitsTrueOther ? this : makeAndSequence(Not.Create(lengthTest), this);
+
+                    static Tests makeOrSequence(Tests t1, Tests.One t2)
+                    {
+                        if (t1 is True) return t1;
```

Comment:
nit: consider extracting those to be factories on `OrSequence` and `AndSequence`. Seems generally useful. If so, may want to special-case `False`/`True` cases, even if not excercised yet.

Reply:
I didn't do that because we'd need to inline the entire Create(ArrayBuilder) logic for a pair of tests for it to be correct:
```cs
return (left, right) switch
{
    (True, _) => True.Instance,
    (_, True) => True.Instance,
    (False, var other) => other,
    (var other, False) => other,
    (OrSequence leftSeq, OrSequence rightSeq) => new OrSequence(leftSeq.RemainingTests.AddRange(rightSeq.RemainingTests)),
    (OrSequence leftSeq, var other) => new OrSequence(leftSeq.RemainingTests.Add(other)),
    (var other, OrSequence rightSeq) => new OrSequence(rightSeq.RemainingTests.Insert(0, other)),
    _ => new OrSequence(ImmutableArray.Create(left, right))
};
```
If that sounds good to you, I can also update MakeTestsAndBindingsForBinaryPattern to use it.

Reply:
Ah okay. That would be too much untested code. Let's not do it.